File: 322-coin-change/322-coin-change.java
class Solution {
    
    int Coin(int[] coins, int amount, int n, int[][] dp)
    {
        for(int i=2; i<=n; i++)
        {
            for(int j=1; j<=amount; j++)
            {
                if(coins[i-1] <= j)
                {
                    dp[i][j] = Math.min(dp[i][j-coins[i-1]]+1, dp[i-1][j]);
                }
                else
                {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }
        if(dp[n][amount] == Integer.MAX_VALUE-1)   return -1;
    return dp[n][amount];
    }
    
    public int coinChange(int[] coins, int amount)
    {
        int n = coins.length;
        int[][] dp = new int[n+1][amount+1];
        
        //initialization of 1st row 1st column
        for(int i=0; i<=n; i++)
        {
            for(int j=0; j<=amount; j++)
            {
                if(j==0)   dp[i][j] = 0;
                if(i==0)   dp[i][j] = Integer.MAX_VALUE-1;
            }
        }
        
        //initialization of 2nd row
        int i=1;
        for(int j=1; j<=amount; j++)
        {
            if(j%coins[0]==0)   dp[i][j] = j/coins[0];
            else                dp[i][j] = Integer.MAX_VALUE-1;
        }
      return Coin(coins, amount, n, dp); 
    }
}

File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int solve(vector<int>nums, int T, int ind, vector<vector<int>>&dp)
    {
        if(ind==0)
        {
            if(T%nums[ind]==0)
            {
                return T/nums[ind];
            }
            else
            {
                return INT_MAX-1;
            }
        }
        
        if(dp[ind][T] != -1)       return dp[ind][T];
        
        int nonPick = solve(nums, T, ind-1, dp);
        int pick = INT_MAX-1;
        
        if(T>=nums[ind])
        {
            T=T-nums[ind];
            pick = 1+solve(nums, T-nums[ind], ind, dp);
        }
        
        return dp[ind][T] = min(pick, nonPick);
    }
    
    int coinChange(vector<int>& nums, int amount)
    {
        if(amount==0) return 0;
        int n = nums.size();
        vector<vector<int>>dp(n, vector<int>(amount+1, 0));
        
        for(int T=0; T<=amount; T++)
        {
            if(T%nums[0]==0)
            {
                dp[0][T] = T/nums[0];
            }
            else
            {
                dp[0][T] = INT_MAX-100;
            }
        }
        
        for(int i=1; i<n; i++)
        {
            for(int T=0; T<=amount; T++)
            {
                int nonPick = dp[i-1][T];
                int pick = INT_MAX-100;

                if(T>=nums[i])
                {
                    pick = 1+dp[i][T-nums[i]];
                }

                dp[i][T] = min(pick, nonPick);
            }
        }
        
        int ans = dp[n-1][amount];
        return ans==INT_MAX-100?-1:ans;
        
         
    }
};